ath79: add support for NEC Aterm WG800HP

musashino205 opened the pull request:
This PR includes following changes:

- fix pinmux reg value for QCA956x SoC
  - qca9557.dtsi has same issue?
- add "hex pattern mode" for xorimage
- add support for NEC Aterm WG800HP

musashino205 committed:
ath79: fix pinmux reg value for QCA956x

The range of pinmux reg property "<0x1804002c 0x40>" for QCA956x SoC
does not includes GPIO_FUNCTION register.

If the device uses "&jtag_disable_pins", this causes the following
errors:

[    1.982937] pinctrl-single 1804002c.pinmux: mux offset out of range: 0x40 (0x40)
[    1.990622] pinctrl-single 1804002c.pinmux: could not add functions for pinmux_jtag_disable_pins 64x

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
firmware-utils: add hex pattern mode for xorimage

This commit adds "hex pattern mode" to xorimage. This mode allows xor
with a hexadecimal pattern that cannot be expressed with ASCII
charactors.

usage (example):
  xorimage -i firmware.bin -o firmware.bin.new -p 6A57190601121E4C -x

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
ath79: add support for NEC Aterm WG800HP

NEC Aterm WG800HP is a 2.4/5 GHz band 11ac router, based on Qualcomm
Atheros QCA9563.

Specification:

- Qualcomm Atheros QCA9563
- 64 MB of RAM (DDR2)
- 8 MB of Flash (SPI-NOR)
- 2.4/5 GHz wifi
  - 2.4 GHz: 2T2R (SoC internal)
  - 5 GHz: 1T1R (QCA9887)
- 4x 10/100/1000 Mbps Ethernet
- 8x LEDs, 3x keys (2x buttons, 1x slide-switch)
- UART through-hole on PCB (J2)
  - Vcc, GND, NC, TX, RX from SoC side
  - 115200n8

Flash instruction using factory image:

1. Connect the computer to the LAN port on WG800HP
2. Connect power cable to WG800HP and turn on it
3. Access to "http://192.168.10.1/" and open firmware update page
("ファームウェア更新")
4. Select the OpenWrt factory image and click update ("更新") button
5. Wait ~150 seconds to complete flashing

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

ynezz commented on target/linux/brcm63xx/image/Makefile:
```diff
@@ -132,11 +132,6 @@ define Build/spw303v-bin
 	mv $@.spw303v $@
 endef
 
-define Build/xor-image
```
@musashino205 This is good idea in itself, but hidden in the patch for new board in `ath79` platform, althought touching code in `brcm63xx` platform. So it probably belongs to separate patch for reusable `xor-image` image target.

chunkeey commented on tools/firmware-utils/src/xorimage.c:
```diff
@@ -24,7 +24,11 @@
 #include <unistd.h>
 #include <sys/stat.h>
 
+#define TRUE 1
```
Why not C99's [stdbool.h](https://en.wikipedia.org/wiki/C_data_types#stdbool.h)?

musashino205 replied:
I do not have much knowledge about C language.
There are two methods of adding define for TRUE/FALSE like *srec2bin.c* and including ```stdbool.h``` like *mktplinkfw.c*, so I did not know which to use.

Should I use stdbool.h?

chunkeey replied:
> Should I use stdbool.h?

I think so, since you are using the "for' loop initial declarations" (a C99 feature) a bit further down in line 118 of the same file.

musashino205 replied:
OK, I'll modify it.

musashino205 committed:
build: move xor-image into image-commands

I moved xor-image into image-commands.mk to use it in ath79 target.

It required for NEC WG800HP.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

chunkeey commented on target/linux/ath79/dts/qca9563_nec_wg800hp.dts:
```diff
@@ -0,0 +1,196 @@
+		wlan2g_red {
+			label = "wg800hp:red:wlan2g";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
The "off" state is already the default. You don't need to explicitly set it.

Though, I'm fine with it. Problem is: Somebody is bound to go on a crusade and remove all "default-state = off".

musashino205 replied:
Oh, I see. I'll drop it.

chunkeey replied:
> Problem is: Somebody is bound to go on a crusade and remove all "default-state = off".

And It just happened. 

https://patchwork.ozlabs.org/patch/1014114/

Good timing.

musashino205 replied:
Oh...

musashino205 commented on target/linux/ath79/dts/qca9563_nec_wg800hp.dts:
```diff
@@ -0,0 +1,196 @@
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		phy-mode = "sgmii";
```
I see. I'll drop it.

musashino205 commented on target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -96,6 +96,10 @@ case "$FIRMWARE" in
 		;;
+	nec,wg800hp)
+		ath10kcal_extract "art" 20480 2116
+		ath10kcal_patch_mac $(mtd_get_mac_text board_data 2176)
```
OK, I'll try it.

musashino205 commented:
Done.

chunkeey commented on tools/firmware-utils/src/xorimage.c:
```diff
@@ -100,6 +107,17 @@ int main(int argc, char **argv)
+
+		for (int i = 0; i < (p_len / 2); i++) {
+			sscanf(pattern + (i * 2), "%2x", &hex_buf);
+			hex_pattern[i] = (char)hex_buf;
```
Because hex_pattern is declared as

      char *hex_pattern = default_pattern; // = "12345678";

the hex_pattern can store at most 8 bytes.

So, this can/will cause a "segmentation fault", if the provided hex pattern is big enough.

    $ ./xorimage -i xorimage -o xorimage-xor -x -p "$(seq 0 8000)"
    Segmentation fault

musashino205 replied:
Hmm...sorry, I confused...

As a pure question (due to ignorance), why is it no problem to pass a 32-byte string to ```pattern``` declared similarly by an 8-byte character string?
In that case, when converted from a character string to a hexadecimal value and passed to ```hex_pattern```, eventually it has a 16-byte value...

musashino205 replied:
Thanks for your detailed advice, it seems I got a rough understanding...
Then, how long should I limit ```hex_pattern``` to?

chunkeey replied:
> Then, how long should I limit hex_pattern to?

There's already kind of a limit. Since the pattern is passed as a parameter to the xorimage program, the pattern has to go through a execve system call which enforces a string length
check of MAX_ARG_STRLEN.

https://elixir.bootlin.com/linux/v4.20-rc7/source/include/uapi/linux/binfmts.h#L15
    /*
     * These are the maximum length and maximum number of strings passed to the
     * execve() system call.  MAX_ARG_STRLEN is essentially random but serves to
     * prevent the kernel from being unduly impacted by misaddressed pointers.
     * MAX_ARG_STRINGS is chosen to fit in a signed 32-bit integer.
     */
    #define MAX_ARG_STRLEN (PAGE_SIZE * 32)

On a 4K PAGE_SIZE System this would be 131072 / 128 KIB.

musashino205 replied:
Hmm...sorry, I searched and investigated about ```execve```, but I did not know how to use ```MAX_ARG_STRLEN``` through it from C source...

chunkeey replied:
> Hmm...sorry, I searched and investigated about `execve`, but I did not know how to use `MAX_ARG_STRLEN` through it from C source...

Hm, I'm getting the feeling that we are unfortunately talking past each other at this point :sob: .

The pattern string can't exceed MAX_ARG_STRLEN by design. So you could just allocate a memory buffer on the heap which holds (strlen(pattern) / 2) bytes and free it after use. But you could also declare something like char hex_pattern[128] buffer and check that (strlen(pattern) / 2) does not exceed the sizeof(hex_pattern). Or you can come up with another option. As long as it passes the fuzzing I do it's "just fine" :+1: .

musashino205 replied:
I see...
Then, how about with the following changes? (Almost as your advice is...)
```char *hex_pattern = default_pattern;``` -> ```char hex_pattern[128];```

and add
```
		if ((p_len / 2) >= sizeof(hex_pattern)) {
			fprintf(stderr, "provided hex pattern is too long\n");
			usage();
		}
```

musashino205 commented:
Fixed xorimage.c and rebased from master.

chunkeey commented:
Merged. Thanks.
